File: hangang_web/lib/Model/busking.dart
// busking.dart
import 'dart:convert';

// Model/busking.dart (핵심만)
class Busking {
  final String? id;      // ✅ Mongo _id (string으로 수신)
  final String userid;
  final String name;
  final String date;
  final String category;
  final String content;
  final String bandName;
  final int state;

  Busking({
    this.id,
    required this.userid,
    required this.name,
    required this.date,
    required this.category,
    required this.content,
    required this.bandName,
    required this.state,
  });

  factory Busking.fromJson(Map<String, dynamic> json) => Busking(
        id: json['_id'] as String?,                 // ✅ _id 매핑
        userid: json['userid'] as String,
        name: json['name'] as String,
        date: json['date'] as String,
        category: json['category'] as String,
        content: json['content'] as String,
        bandName: json['bandName'] as String,
        state: (json['state'] as num).toInt(),
      );

  Map<String, dynamic> toJson() => {
        // insert 시 _id는 서버가 생성
        'userid': userid,
        'name': name,
        'date': date,
        'category': category,
        'content': content,
        'bandName': bandName,
        'state': state,
      };

  Busking copyWith({
    String? id,
    String? userid,
    String? name,
    String? date,
    String? category,
    String? content,
    String? bandName,
    int? state,
  }) {
    return Busking(
      id: id ?? this.id,
      userid: userid ?? this.userid,
      name: name ?? this.name,
      date: date ?? this.date,
      category: category ?? this.category,
      content: content ?? this.content,
      bandName: bandName ?? this.bandName,
      state: state ?? this.state,
    );
  }
}

File: Python/routes/busking.py
# routes/busking.py

from fastapi import APIRouter, HTTPException
from motor.motor_asyncio import AsyncIOMotorClient
from pydantic import BaseModel
from typing import Optional
from bson import ObjectId 
import base64
import os

# Router 생성
router = APIRouter()

# MongoDB 연결
MONGO_URI = os.getenv("MONGODB_URI", "mongodb+srv://rlaxoals990504:team2public@team2.jozn4ix.mongodb.net/")
client = AsyncIOMotorClient(MONGO_URI)
db = client.lecture
collection_busking = db.busking

# Pydantic 모델
class busking(BaseModel):
    userid : str
    name : str
    date : str
    category : str
    content : str
    bandName : str
    state : int
    # image: Optional[str] = None  # base64 문자열

class buskingUpdate(BaseModel):
    userid : Optional[str] = None
    name : Optional[str] = None
    date : Optional[str] = None
    category : Optional[str] = None
    content : Optional[str] = None
    bandName : Optional[str] = None
    state : Optional[int] = None

# 유틸: Mongo 문서 포맷 보정
def normalize_busking(doc: dict) -> dict:
    if not doc:
        return doc
    doc["_id"] = str(doc.get("_id"))  # ✅ ObjectId -> str
    return doc

def to_object_id(id_str: str) -> ObjectId:
    try:
        return ObjectId(id_str)
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid id format")



@router.get('/busking/select')
async def select():
    busking = await collection_busking.find().to_list(None)
    results = [normalize_busking(s) for s in busking]
    return {'results': results}

@router.post('/busking/insert')
async def insert(busking: busking):
    # userid 중복 검사
    # existing = await collection_busking.find_one({'userid': busking.userid})
    # if existing:
    #     raise HTTPException(status_code=400, detail='busking is existed.')

    data = busking.dict()
    # image(base64) → bytes
    if data.get('image'):
        try:
            data['image'] = base64.b64decode(data['image'])
        except Exception:
            raise HTTPException(status_code=400, detail='Invalid Base64 image')

    await collection_busking.insert_one(data)
    return {'result': 'OK'}

@router.put("/busking/update/{id}")
async def update(id: str, payload: buskingUpdate):
    data = payload.dict(exclude_unset=True)
    if not data:
        raise HTTPException(status_code=400, detail="No Field For Update")

    oid = to_object_id(id)  # ✅ 문자열 → ObjectId
    result = await collection_busking.update_one({"_id": oid}, {"$set": data})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="busking Not Found")
    return {"result": "OK"}

@router.delete("/busking/delete/{id}")
async def delete(id: str):
    oid = to_object_id(id)  # ✅ 문자열 → ObjectId
    result = await collection_busking.delete_one({"_id": oid})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="busking Not Found")
    return {"result": "OK"}